### ISSUE TITLE
Attribute error: no attribute '__reduce_cython__'

### ISSUE THREAD

**Issue by SophiaNasr** [pynbody/pynbody#503]
Hello,

I've installed pynbody on my virtualbox in Ubuntu 18.04.2, and python 3.7.1. When I try to 

import pynbody
import matplotlib.pylab as plt 

I get the following error:

```pytb
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-2-0c12c4748636> in <module>
----> 1 import pynbody
      2 import matplotlib.pylab as plt
      3 

~/anaconda3/lib/python3.7/site-packages/pynbody/__init__.py in <module>
     19 from .configuration import config, config_parser, logger
     20 
---> 21 from . import util, filt, array, family, snapshot
     22 from .snapshot import tipsy, gadget, gadgethdf, ramses, grafic, nchilada, ascii
     23 from . import analysis, halo, derived, bridge, gravity, sph, transformation

~/anaconda3/lib/python3.7/site-packages/pynbody/util.py in <module>
     27 
     28 logger = logging.getLogger('pynbody.util')
---> 29 from ._util import *
     30 
     31 

~/anaconda3/lib/python3.7/site-packages/pynbody/_util.cpython-37m-x86_64-linux-gnu.so in init pynbody._util()

AttributeError: type object 'pynbody._util.array' has no attribute '__reduce_cython__'
```
The '__reduce_cython__' error is something that comes up a few times in Google searches, but I never find something that looks useful to me, as the results usually include other problems, and responses are mostly to those problems, and not the error in question. Would anyone know whether simply removing cython would solve the problem?? Or would I need to do something different? 

Thanks so much for your help.

**Comment by apontzen** [comment]
Looks from google like this is a bug  with specific versions of cython. A few questions: did you install pynbody from git or from a prepackaged version? What version of cython is installed on your system?

Anyway, I’d definitely recommend trying updating cython and then installing the latest pynbody from git.

**Comment by SophiaNasr** [comment]
> Looks from google like this is a bug with specific versions of cython. A few questions: did you install pynbody from git or from a prepackaged version? What version of cython is installed on your system?
> 
> Anyway, I’d definitely recommend trying updating cython and then installing the latest pynbody from git.

Hi Andrew, thank you very much for your response. I updated cython with conda update cython (it is now version 0.29.4), and then downloaded pynbody from git (using pip install git+git://github.com/pynbody/pynbody.git) . After doing that, the previous error went away, and this one came up:

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-0c12c4748636> in <module>
----> 1 import pynbody
      2 import matplotlib.pylab as plt
      3 

~/anaconda3/lib/python3.7/site-packages/pynbody/__init__.py in <module>
     19 from .configuration import config, config_parser, logger
     20 
---> 21 from . import util, filt, array, family, snapshot
     22 from .snapshot import tipsy, gadget, gadgethdf, ramses, grafic, nchilada, ascii
     23 from . import analysis, halo, derived, bridge, gravity, sph, transformation

~/anaconda3/lib/python3.7/site-packages/pynbody/util.py in <module>
     27 
     28 logger = logging.getLogger('pynbody.util')
---> 29 from ._util import *
     30 
     31 

__init__.pxd in init pynbody._util()

ValueError: numpy.ufunc size changed, may indicate binary incompatibility. Expected 216 from C header, got 192 from PyObject

_____________________________________________

After this, I Googled this error and one thing that came up said numpy should be updated, so I decided to conda update all just to make sure all packages are ready. Then I got the same original error (no attribute reduce cython) so I restarted the kernel and ran the imports again, and now the above numpy error is back. 

One thing I should point out is that I had python 3 installed before with anaconda 3, but a research project required python 2, so I removed python 3. Now, I had to install it again, so I removed python 2 and installed python with anaconda (now I have both), and the first time I did that I had trouble with residual files so anaconda wouldn't install packages. I removed anaconda, cleaned out the bash file, and installed once more, and everything worked and I was able to install all I have thus far. I think I did this properly because everything else works, the only thing that does not is the new pynbody package, so I am guessing there is some other thing missing. Would there be anything else I could share that might help you help me?

**Comment by SophiaNasr** [comment]
Update: the issue was that when I updated cython, I did it using conda update. I redid it by pip install --upgrading, then I installed pynbody from GitHub, and finally, pip install --upgrade numpy. Now I can import the package without any problems. I guess the bottom line is that updating with pip works better than with conda.

**Comment by apontzen** [comment]
Glad to hear this was resolved. It's very frustrating when you end up with binary incompatibilities like this. Basically it looks like the original pynbody install somehow compiled against the wrong version of numpy, but I don't really understand how.

Do you know the exact previous version of cython you were using? It might be worth putting in a warning to our install script if that version is detected, since this seems to be a known issue.

**Comment by SophiaNasr** [comment]
> Glad to hear this was resolved. It's very frustrating when you end up with binary incompatibilities like this. Basically it looks like the original pynbody install somehow compiled against the wrong version of numpy, but I don't really understand how.
> 
> Do you know the exact previous version of cython you were using? It might be worth putting in a warning to our install script if that version is detected, since this seems to be a known issue.

The previous cython version I had was 0.29.4. The updated version is 0.29.5. It is also worth noting, though, that it isn't until after updated (via pip) numpy, after updating cython, that the error went away. The previous version of numpy was 1.15.4, and my current is 1.16.1.

**Comment by DhirajReddy** [comment]
> Update: the issue was that when I updated cython, I did it using conda update. I redid it by pip install --upgrading, then I installed pynbody from GitHub, and finally, pip install --upgrade numpy. Now I can import the package without any problems. I guess the bottom line is that updating with pip works better than with conda.

I was facing the same problem when I installed wordbatch, The issue was resolved after I ready this comment and updated my numpy. Thank you very much

**Comment by texadactyl** [comment]
Just a note about the ticking clock on Python 2 - In 2020, Python 2 will be deemed obsolete by the (unpaid) Python support folks.  Consider conversion projects when you get a chance.  Spyder is a great visual-oriented IDE to use for converting and finding other warnings that could become serious in the future.

I have found that 99% of my own incompatibility issues centered around use of ```print```.
Python 2: ```print a, b```
Python 2 & 3: ```print(a, b)```
That simple.

**Comment by yustiks** [comment]
solved the problem by installing numpy:
pip install numpy==1.16.1

**Comment by 3180100753** [comment]
thanks!

**Comment by Pooret** [comment]
Having this problem now with gensim

`AttributeError: type object 'gensim._matutils.array' has no attribute '__reduce_cython__'
`
Prior to this, had this error...

`ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject`

Fixed this by upgrading numpy to 1.21.0, but above issue now arises and haven't found a solution as of yet. I've upgraded cython, gensim, scipy to current versions.
